tests/unit/testsettings.cpp

```diff
@@ -0,0 +1,101 @@
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "testsettings.h"
+#include "../../src/settingsholder.h"
+#include "helper.h"
+
+void TestSettings::transactionErrors() {
+  SettingsHolder settingsHolder;
+
+  QVERIFY(!settingsHolder.commitTransaction());
+  QVERIFY(!settingsHolder.rollbackTransaction());
+
+  QVERIFY(settingsHolder.beginTransaction());
+  QVERIFY(!settingsHolder.beginTransaction());
+  QVERIFY(settingsHolder.commitTransaction());
+  QVERIFY(!settingsHolder.commitTransaction());
+
+  QVERIFY(settingsHolder.beginTransaction());
+  QVERIFY(!settingsHolder.beginTransaction());
+  QVERIFY(settingsHolder.commitTransaction());
+  QVERIFY(!settingsHolder.rollbackTransaction());
+
+  QVERIFY(settingsHolder.beginTransaction());
+  QVERIFY(!settingsHolder.beginTransaction());
+  QVERIFY(settingsHolder.rollbackTransaction());
+  QVERIFY(!settingsHolder.rollbackTransaction());
+
+  QVERIFY(settingsHolder.beginTransaction());
+  QVERIFY(!settingsHolder.beginTransaction());
+  QVERIFY(settingsHolder.rollbackTransaction());
+  QVERIFY(!settingsHolder.commitTransaction());
+}
+
+void TestSettings::transactionCommit() {
+  SettingsHolder settingsHolder;
+
+  QVERIFY(settingsHolder.beginTransaction());
+  settingsHolder.setTheme("AAA");
+  QCOMPARE(settingsHolder.theme(), "AAA");
+  QVERIFY(settingsHolder.commitTransaction());
+  QCOMPARE(settingsHolder.theme(), "AAA");
+}
+
+void TestSettings::transactionRollback() {
+  SettingsHolder settingsHolder;
+
+  QVERIFY(settingsHolder.beginTransaction());
+
+  settingsHolder.setToken("TOKEN 1");
+  QCOMPARE(settingsHolder.token(), "TOKEN 1");
+
+  settingsHolder.setTheme("AAA");
+  settingsHolder.setTheme("BBB");
+  settingsHolder.setTheme("CCC");
+  QCOMPARE(settingsHolder.theme(), "CCC");
+
+  int count = 0;
+  connect(&settingsHolder, &SettingsHolder::themeChanged, [&]() { ++count; });
+
+  QVERIFY(settingsHolder.rollbackTransaction());
+
+  QCOMPARE(settingsHolder.theme(), DEFAULT_THEME);
+  QCOMPARE(settingsHolder.token(), "TOKEN 1");
+  QCOMPARE(count, 1);
+}
+
+void TestSettings::transactionRollbackStartup() {
+  {
+    SettingsHolder settingsHolder;
+    settingsHolder.doNotClearOnDTOR();
+
+    settingsHolder.setTheme("AAA");
+    QCOMPARE(settingsHolder.theme(), "AAA");
+    settingsHolder.setToken("TOKEN 1");
+    QCOMPARE(settingsHolder.token(), "TOKEN 1");
+  }
+
+  {
+    SettingsHolder settingsHolder;
+    settingsHolder.doNotClearOnDTOR();
+
+    QCOMPARE(settingsHolder.theme(), "AAA");
+    QCOMPARE(settingsHolder.token(), "TOKEN 1");
+
+    QVERIFY(settingsHolder.beginTransaction());
+    settingsHolder.setTheme("BBB");
+    QCOMPARE(settingsHolder.theme(), "BBB");
+    settingsHolder.setToken("TOKEN 2");
+    QCOMPARE(settingsHolder.token(), "TOKEN 2");
+  }
+
+  {
+    SettingsHolder settingsHolder;
+    QCOMPARE(settingsHolder.theme(), "AAA");
```

Comment:
Ha! I'm confused. Why does theme not change?

Reply:
because we are simulating a recovering from a missing rollback/commit

Reply:
because token is not an user-setting

Reply:
Ah, of course. Ok makes sense. Would you mind adding some comments just to make the test easier to read?